### PR TITLE
Add validation for push endpoint connectivity 

### PR DESCRIPTION
**Pull request by subintp**
- Check push endpoint is reachable while creating the subscription

### CODE REVIEW

File: internal/subscription/validation.go [hunk 1]
```diff
@@ -370,6 +383,22 @@ func validateDeadLetterPolicy(_ context.Context, m *Model, req *metrov1.Subscrip
 	return nil
 }
 
+func validatePushEndpoint(_ context.Context, req *metrov1.Subscription, client *http.Client) error {
+	if req.GetPushConfig() != nil {
+		resp, err := client.Get(req.GetPushConfig().PushEndpoint)
```
[the remark this answers is not in the document]
Comment (vnktram) [reply internal/subscription/validation.go:388]: @subintp GET might return an error even if the endpoint is reachable. We want to use DialTimeout instead.

**Review by vnktram — CHANGES_REQUESTED** [review]

File: internal/subscription/validation.go [hunk 2]
```diff
@@ -370,6 +383,24 @@ func validateDeadLetterPolicy(_ context.Context, m *Model, req *metrov1.Subscrip
...
 }
 
+func validatePushEndpoint(_ context.Context, req *metrov1.Subscription) error {
+	if req.GetPushConfig() != nil {
+		u, err := url.Parse(req.GetPushConfig().PushEndpoint)
+		if err != nil {
+			return ErrInvalidPushEndpointURL
+		}
+
+		// eg: DialTimeout("tcp", "golang.org:http", timeout)
```
[the remark this answers is not in the document]
Comment (subintp) [reply internal/subscription/validation.go:393]: Intentionally kept it there to understand the why behind the network address creation

**Comment by sakshi-goyal-razorpay** [comment]
LGTM, the code coverage checks need to be taken care of.

**Review by akashgoyal96 — APPROVED** [review]
LGTM

File: internal/subscription/validation_test.go [hunk 3]
```diff
@@ -445,3 +445,32 @@ func Test_validateFilterExpression(t *testing.T) {
...
+
+func Test_validatePushEndpoint(t *testing.T) {
+	ctx := context.Background()
+
+	type testCase struct {
+		inputConfig *metrov1.PushConfig
+		err         error
+	}
+
+	tests := [2]testCase{
```
Comment (akashgoyal96) [inline internal/subscription/validation_test.go:457]: @subintp can we add another test case to test the InvalidPushEndpointURL error?

**Review by sakshi-goyal-razorpay — APPROVED** [review]

File: internal/subscription/validation.go [hunk 4]
```diff
@@ -370,6 +383,24 @@ func validateDeadLetterPolicy(_ context.Context, m *Model, req *metrov1.Subscrip
...
+		networkAddress := u.Hostname() + ":" + u.Scheme
```
Comment (vnktram) [inline internal/subscription/validation.go:394]: Shouldn't the format be `Scheme : Hostname` ?

Comment (vnktram) [reply internal/subscription/validation.go:394]: Say `https://dummy.rzp.in` ?

**Review by vnktram — CHANGES_REQUESTED** [review]

File: internal/subscription/validation.go (hunk 4 above)
Comment (subintp) [reply internal/subscription/validation.go:394]: No the other way 
```

/ Examples:
//	Dial("tcp", "golang.org:http")
//	Dial("tcp", "192.0.2.1:http")
//	Dial("tcp", "198.51.100.1:80")
//	Dial("udp", "[2001:db8::1]:domain")
//	Dial("udp", "[fe80::1%lo0]:53")
//	Dial("tcp", ":80")
```

File: internal/subscription/validation_test.go (hunk 3 above)
Comment (subintp) [reply internal/subscription/validation_test.go]: Done

**Review by vnktram — APPROVED** [review]
LGTM. Congrats on your first Metro PR @subintp

**Review by akashgoyal96 — APPROVED** [review]
LGTM

**Comment by subintp** [comment]
Tested on stage. Merging the PR
Screenshot 2022-08-18 at 10 58 29 PM

**Pull request merged**